core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2921,6 +2925,23 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
   )(f: (Stream[F, O], Stream[F2, O2]) => Stream[F2, O3]): Stream[F2, O3] =
     f(this, s2)
 
+  /** Transforms this stream, explicitly extending the current scope through the given pipe.
+    *
+    * Use this when implementing a pipe where the resulting stream is not directly constructed from
+    * the source stream, e.g. when sending the source stream through a Channel and returning the
+    * channel's stream.
+    */
+  def extendScopeThrough[F2[x] >: F[x], O2](
+      f: Stream[F, O] => Stream[F2, O2]
+  )(implicit F: MonadError[F2, Throwable]): Stream[F2, O2] =
+    this.pull.peek
```

Comment:
Is it guaranteed to be safe (in the context of scopes) to use `.peek` before `Pull.extendScopeTo`?

Reply:
I'm assuming scopes work properly for simple streams. 🤞 Do you have a specific scenario in mind?

The fundamental issue is that I need to get ahold of a scope before we go adding more finalizers to `source` to avoid deadlocking in parEvalMap and I don't see a way to do that without `uncons`. `peek` is just a convenience helper.

Reply:
> Do you have a specific scenario in mind?

No, not really, it's just me being paranoid :)

I don't see a way to do it without `uncons` either, I was thinking about swapping the order of those `Pull`s:
```scala
def extendScopeThrough[F2[x] >: F[x], O2](
      f: Stream[F, O] => Stream[F2, O2]
  )(implicit F: MonadError[F2, Throwable]): Stream[F2, O2] =
    Pull
      .extendScopeTo(this.covary[F2])
      .flatMap { stream =>
        stream.pull.peek.flatMap {
          case Some((_, tl)) => f(tl).underlying // <---------------
          case None          => f(Stream.empty).underlying
        }
      }
      .stream
```

but the problem is `tl` is not a `Stream[F, O]` (it's a `Stream[F2, O]`), and `.covary[F2]` at the beginning has to be there because we only have `MonadError` for `F2`.

Reply:
Just make it `f: Stream[F2, O] => Stream[F2, O2]`. Requires type annotations at the call sites, but otherwise, this version also passes the tests.

Unless we can find an observable difference I'd rather keep the current version to avoid the extra annotations.

Reply:
I agree :)